File: src/shared/api/protocol.ts
export const PROTOCOL_VERSION = 2 as const;

export interface PlaybackState {
  positionSeconds: number;
  playing: boolean;
  playbackRate: number;
  updatedAtMs: number;
  revision: number;
}

export interface Participant {
  participantId: string;
  displayName: string;
  isHost: boolean;
  pingMs: number | null;
}

export interface ExternalSubtitle {
  id: string;
  name: string;
  language: string | null;
}

export interface PlaylistItem {
  id: string;
  name: string;
  progressSeconds: number;
  durationSeconds: number;
  externalSubtitles: ExternalSubtitle[];
}

export interface RoomSnapshot {
  roomCode: string;
  mediaName: string;
  playlist: PlaylistItem[];
  activePlaylistItemId: string | null;
  participantCount: number;
  participants?: Participant[];
  hostParticipantId: string;
  mediaAccessToken: string;
  playback: PlaybackState;
}

export type PlaybackAction = "play" | "pause" | "seek";

export type ClientMessage =
  | {
      type: "create_room";
      payload: {
        version: typeof PROTOCOL_VERSION;
        roomCode: string;
        mediaName: string;
        playlist: PlaylistItem[];
        activePlaylistItemId: string | null;
        participantId: string;
        displayName: string;
        mediaAccessToken: string;
      };
    }
  | {
      type: "join_room";
      payload: {
        version: typeof PROTOCOL_VERSION;
        roomCode: string;
        participantId: string;
        displayName: string;
      };
    }
  | {
      type: "playback_command";
      payload: {
        roomCode: string;
        action: PlaybackAction;
        positionSeconds: number;
      };
    }
  | {
      type: "playback_rate";
      payload: {
        roomCode: string;
        playbackRate: number;
        positionSeconds: number;
      };
    }
  | {
      type: "playlist_update";
      payload: {
        roomCode: string;
        playlist: PlaylistItem[];
        activePlaylistItemId: string | null;
      };
    }
  | { type: "ping"; payload: { clientTimeMs: number } }
  | { type: "latency_report"; payload: { pingMs: number } };

export type ServerMessage =
  | { type: "room_snapshot"; payload: { room: RoomSnapshot } }
  | {
      type: "participant_count";
      payload: { roomCode: string; participantCount: number };
    }
  | {
      type: "participant_joined";
      payload: { roomCode: string; participantId: string; displayName: string };
    }
  | {
      type: "participant_list";
      payload: { roomCode: string; participants: Participant[] };
    }
  | {
      type: "playback_state";
      payload: { roomCode: string; playback: PlaybackState };
    }
  | {
      type: "playlist_state";
      payload: {
        roomCode: string;
        playlist: PlaylistItem[];
        activePlaylistItemId: string | null;
      };
    }
  | {
      type: "room_closed";
      payload: { roomCode: string; reason: "host_left" | string };
    }
  | {
      type: "pong";
      payload: { clientTimeMs: number; serverTimeMs: number };
    }
  | { type: "error"; payload: { code: string; message: string } };

export function parseServerMessage(value: string): ServerMessage | null {
  try {
    const parsed: unknown = JSON.parse(value);
    if (!parsed || typeof parsed !== "object" || !("type" in parsed) || !("payload" in parsed)) {
      return null;
    }

    const type = (parsed as { type: unknown }).type;
    if (
      type !== "room_snapshot" &&
      type !== "participant_count" &&
      type !== "participant_joined" &&
      type !== "participant_list" &&
      type !== "playback_state" &&
      type !== "playlist_state" &&
      type !== "room_closed" &&
      type !== "pong" &&
      type !== "error"
    ) {
      return null;
    }

    return parsed as ServerMessage;
  } catch {
    return null;
  }
}

File: src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u8 = 2;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaybackState {
    pub position_seconds: f64,
    pub playing: bool,
    pub playback_rate: f64,
    pub updated_at_ms: u64,
    pub revision: u64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ParticipantSnapshot {
    pub participant_id: String,
    pub display_name: String,
    pub is_host: bool,
    pub ping_ms: Option<u64>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExternalSubtitle {
    pub id: String,
    pub name: String,
    pub language: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaylistItem {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub progress_seconds: f64,
    #[serde(default)]
    pub duration_seconds: f64,
    #[serde(default)]
    pub external_subtitles: Vec<ExternalSubtitle>,
}

impl Default for PlaybackState {
    fn default() -> Self {
        Self {
            position_seconds: 0.0,
            playing: false,
            playback_rate: 1.0,
            updated_at_ms: 0,
            revision: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RoomSnapshot {
    pub room_code: String,
    pub media_name: String,
    pub playlist: Vec<PlaylistItem>,
    pub active_playlist_item_id: Option<String>,
    pub participant_count: usize,
    pub participants: Vec<ParticipantSnapshot>,
    pub host_participant_id: String,
    pub media_access_token: String,
    pub playback: PlaybackState,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum ClientMessage {
    CreateRoom(CreateRoomPayload),
    JoinRoom(JoinRoomPayload),
    PlaybackCommand(PlaybackCommandPayload),
    PlaybackRate(PlaybackRatePayload),
    PlaylistUpdate(PlaylistUpdatePayload),
    Ping(PingPayload),
    LatencyReport(LatencyReportPayload),
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRoomPayload {
    pub version: u8,
    pub room_code: String,
    pub media_name: String,
    #[serde(default)]
    pub playlist: Vec<PlaylistItem>,
    #[serde(default)]
    pub active_playlist_item_id: Option<String>,
    pub participant_id: String,
    #[allow(dead_code)]
    pub display_name: String,
    pub media_access_token: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinRoomPayload {
    pub version: u8,
    pub room_code: String,
    pub participant_id: String,
    #[allow(dead_code)]
    pub display_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaybackCommandPayload {
    pub room_code: String,
    pub action: PlaybackAction,
    pub position_seconds: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaybackRatePayload {
    pub room_code: String,
    pub playback_rate: f64,
    pub position_seconds: f64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaylistUpdatePayload {
    pub room_code: String,
    pub playlist: Vec<PlaylistItem>,
    pub active_playlist_item_id: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlaybackAction {
    Play,
    Pause,
    Seek,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PingPayload {
    pub client_time_ms: u64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LatencyReportPayload {
    pub ping_ms: u64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
pub enum ServerMessage {
    RoomSnapshot {
        room: RoomSnapshot,
    },
    ParticipantCount {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "participantCount")]
        participant_count: usize,
    },
    ParticipantJoined {
        #[serde(rename = "roomCode")]
        room_code: String,
        #[serde(rename = "participantId")]
        participant_id: String,
        #[serde(rename = "displayName")]
        display_name: String,
    },
    ParticipantList {
        #[serde(rename = "roomCode")]
        room_code: String,
        participants: Vec<ParticipantSnapshot>,
    },
    PlaybackState {
        #[serde(rename = "roomCode")]
        room_code: String,
        playback: PlaybackState,
    },
    PlaylistState {
        #[serde(rename = "roomCode")]
        room_code: String,
        playlist: Vec<PlaylistItem>,
        #[serde(rename = "activePlaylistItemId")]
        active_playlist_item_id: Option<String>,
    },
    RoomClosed {
        #[serde(rename = "roomCode")]
        room_code: String,
        reason: String,
    },
    Pong {
        #[serde(rename = "clientTimeMs")]
        client_time_ms: u64,
        #[serde(rename = "serverTimeMs")]
        server_time_ms: u64,
    },
    Error {
        code: String,
        message: String,
    },
}

impl ServerMessage {
    pub fn error(code: impl Into<String>, message: impl Into<String>) -> Self {
        Self::Error {
            code: code.into(),
            message: message.into(),
        }
    }
}
